### PR TITLE
feat: Add support for CAPX

### PR DESCRIPTION
**Pull request by deepakm-ntnx**
pushed from https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/pull/384 to this repo
Also patches from  https://github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/pull/5 are required

to get latest nutanix apis from capx repo
<pre>
make apis.sync
</pre>

to regenerate example Cluster with topology yamls
<pre>
 make examples.sync
</pre>

To run unit tests
<pre>
make lint
 make test
</pre>

To deploy cre controller as well as nutanix provider and cluster class of nutanix
<pre>
make dev.run-on-kind 
make make dev.update-webhook-image-on-kind
</pre>

To deploy nutanix cluster with topology with cre annotations 
<pre>
export KUBECONFIG=/go/src/github.com/deepakm-ntnx/capi-runtime-extensions/.local/kind/capi-runtime-extensions-dev/kubeconfig
clusterctl generate cluster nutanix-quick-start-helm-addon-cilium \
  --from examples/capi-quick-start/nutanix-cluster-cilium-helm-addon.yaml \
  --kubernetes-version v1.29.1 \
  --worker-machine-count 1 | \
  kubectl apply --server-side -f -
</pre>

### CODE REVIEW

File: charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+  template:
+    spec:
+      joinConfiguration:
+        nodeRegistration:
+          kubeletExtraArgs:
+            cloud-provider: external
+            eviction-hard: nodefs.available<10%,nodefs.inodesFree<5%,imagefs.available<15%,memory.available<100Mi,imagefs.inodesFree<10%
+            tls-cipher-suites: TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
+      postKubeadmCommands:
+      - echo "after kubeadm call" > /var/log/postkubeadm.log
```
Comment (faiq) [inline charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:17]: Why is this necessary?

File: charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+          controllerManager:
+            extraArgs:
+              cloud-provider: external
+              tls-cipher-suites: TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
+          scheduler:
+            extraArgs:
+              tls-cipher-suites: TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
+        files:
+        - content: |
+            apiVersion: v1
```
Comment (faiq) [inline charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:141]: i'd like to see this modified with a patch handler, not like this.

File: charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+spec:
+  template:
+    spec:
+      kubeadmConfigSpec:
+        clusterConfiguration:
+          apiServer:
+            certSANs:
+            - localhost
+            - 127.0.0.1
+            - 0.0.0.0
```
Comment (faiq) [inline charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:128]: This should not be in the base template.

Comment (faiq) [reply charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:128]: https://github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/blob/main/docs/content/customization/generic/extra-apiserver-cert-sans.md

this is how we can set it through the cluster variable, and our default Cluster implementation can have this. Or we can add this in automatically via code.

File: charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml (hunk 2 above)
Comment (deepakm-ntnx) [reply charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:141]: https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix/pull/392#pullrequestreview-1949576783. it was discussed earlier and its good to make everything variable, it was deemed not urgent. you can file a tracker if you think its a must fix with reason at this time

File: charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml (hunk 3 above)
Comment (deepakm-ntnx) [reply charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:128]: we already have PR open https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix/pull/372 for capx.

File: charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml (hunk 1 above)
Comment (deepakm-ntnx) [reply charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:17]: which part? echo stmt?

Comment (faiq) [reply charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:17]: Yes the echo

File: charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml (hunk 3 above)
Comment (faiq) [reply charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:128]: This shouldn't be a json patch and should instead be a done with GeneratePatches

File: charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml (hunk 1 above)
Comment (deepakm-ntnx) [reply charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:17]: might have been added long back for checkpointing for debugging purposes only. i would keep it and if needed remove it in another PR

File: charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml (hunk 2 above)
Comment (faiq) [reply charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:141]: I don't think this needs to be handled by a variable, but we can reuse this code if it's implemented as a patch handler.

File: charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml (hunk 3 above)
Comment (faiq) [reply charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:128]: let's keep these open for more discussion.

Comment (faiq) [reply charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:128]: Just one more point, this clusterclass isn't meant to be used by CAPX. It's an opinionated clusterclass for CAREN. We don't need to worry about the other consumers here.

Comment (deepakm-ntnx) [reply charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:128]: I think plan is clear:
- add variable in capx https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix/pull/372
- in subsequent PR use generic https://d2iq-labs.github.io/cluster-api-runtime-extensions-nutanix/customization/generic/extra-apiserver-cert-sans/ to update certSAN 

Is there anything open in your mind?

File: charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml (hunk 2 above)
Comment (jimmidyson) [reply charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:141]: Can we deploy kube-vip as an addon?

Comment (faiq) [reply charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:141]: I don't think so. I'm not sure how a static pod deployment like kube-vip would get deployed as an addon.

It seems like something we want as a static pod

Comment (jimmidyson) [reply charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:141]: Ah good point. But we could do this as a before create cluster hook that adds this static pod in to kubeadm config template, right? This is a generic pod for other providers too, right?

Comment (faiq) [reply charts/cluster-api-runtime-extensions-nutanix/defaultclusterclasses/nutanix-cluster-class.yaml:141]: Yeah, we want to reuse this for providers like vsphere and for preprovisioned. We could do it as a BeforeClusterCreate hook or add it in as GeneratePatch

File: api/v1alpha1/nutanix_clusterconfig_types.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+					Type:        "boolean",
+				},
+				"additionalTrustBundle": {
+					Description: "Name of configMap with certificate trust bundle used for Prism Central connection",
+					Type:        "string",
+				},
+				"credentialSecret": {
+					Description: "Name of a Credential information secret for the target Prism instance",
+					Type:        "string",
+				},
```
Comment (faiq) [inline api/v1alpha1/nutanix_clusterconfig_types.go:63]: Should we use a proper secretRef type?

File: api/v1alpha1/nutanix_clusterconfig_types.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+	// cluster is to identify the cluster (the Prism Element under management of the Prism Central),
+	// in which the Machine's VM will be created. The cluster identifier (uuid or name) can be obtained
+	// from the Prism Central console or using the prism_central API.
+	Cluster NutanixResourceIdentifier `json:"cluster"`
+
+	// subnets holds a list of identifiers (one or more) of the cluster's network subnets
+	// for the Machine's VM to connect to. The subnet identifiers (uuid or name) can be
+	// obtained from the Prism Central console or using the prism_central API.
+	Subnets []NutanixResourceIdentifier `json:"subnets"`
```
Comment (deepakm-ntnx) [inline api/v1alpha1/nutanix_clusterconfig_types.go:123]: currently facing following issue in capi while applying patches <pre>
E0329 21:46:37.778027       1 reconcile_state.go:1113] "Failed to create object" err="NutanixMachineTemplate.infrastructure.cluster.x-k8s.io \"nutanix-quick-start-helm-addon-cilium-npscb\" is invalid: spec.template.spec.subnet[0].type: Unsupported value: \"\": supported values: \"uuid\", \"name\"" controller="topology/cluster" controllerGroup="cluster.x-k8s.io" controllerKind="Cluster" Cluster="default/nutanix-quick-start-helm-addon-cilium" namespace="default" name="nutanix-quick-start-helm-addon-cilium" reconcileID="ded8efe6-16c4-4f6e-a6c7-35c8f6e231c5" resource={"group":"infrastructure.cluster.x-k8s.io","version":"v1beta1","resource":"NutanixMachineTemplate"} NutanixMachineTemplate="default/nutanix-quick-start-helm-addon-cilium-npscb" NutanixMachineTemplate="default/nutanix-quick-start-helm-addon-cilium-npscb"
E0329 21:46:37.779396       1 controller.go:329] "Reconciler error" err="error reconciling the Cluster topology: failed to create NutanixMachineTemplate.infrastructure.cluster.x-k8s.io: FieldValueNotSupported: spec.template.spec.subnet[0].type: Unsupported value: \"\": supported values: \"uuid\", \"name\"" controller="topology/cluster" controllerGroup="cluster.x-k8s.io" controllerKind="Cluster" Cluster="default/nutanix-quick-start-helm-addon-cilium" namespace="default" name="nutanix-quick-start-helm-addon-cilium" reconcileID="ded8efe6-16c4-4f6e-a6c7-35c8f6e231c5"
</pre>

**Comment by dkoshkin** [comment]
@deepakm-ntnx the commits need to have the conventional commits format `fix/feat/build/refactor` etc.

Can you please squash the commits here that make sense to be together, I think that will also make it simpler to review.

File: pkg/handlers/nutanix/mutation/machinedetails/inject_worker.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+			if nutanixMachineDetailsVar.Image.Type == v1alpha1.NutanixIdentifierName {
+				obj.Spec.Template.Spec.Image.Name = nutanixMachineDetailsVar.Image.Name
+			} else {
+				obj.Spec.Template.Spec.Image.UUID = nutanixMachineDetailsVar.Image.UUID
+			}
+
+			obj.Spec.Template.Spec.VCPUSockets = nutanixMachineDetailsVar.VCPUSockets
+			obj.Spec.Template.Spec.VCPUsPerSocket = nutanixMachineDetailsVar.VCPUsPerSocket
+			obj.Spec.Template.Spec.MemorySize = resource.MustParse(
+				nutanixMachineDetailsVar.MemorySize,
```
Comment (dkoshkin) [inline pkg/handlers/nutanix/mutation/machinedetails/inject_worker.go:119]: Maybe we should use `resource.Quantity` in the CAREN API? I can pull it into https://github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/pull/13

File: pkg/handlers/nutanix/mutation/machinedetails/inject_worker.go [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+					subnet.Name = subnetCRE.Name
+				} else {
+					subnet.Type = capxv1.NutanixIdentifierUUID
+					if subnetCRE.UUID == nil || *subnetCRE.UUID == "" {
+						continue
+					}
+					subnet.UUID = subnetCRE.UUID
+				}
+				obj.Spec.Template.Spec.Subnets = append(obj.Spec.Template.Spec.Subnets, subnet)
+			}
```
Comment (dkoshkin) [inline pkg/handlers/nutanix/mutation/machinedetails/inject_worker.go:144]: This will be simplified when we reuse the same types, but lets avoid using "CRE" in the types (since this is already outdated after the rename)
Maybe something like this?
```suggestion
			for _, subnet := range nutanixMachineDetailsVar.Subnets {
				if subnet.Type == "" {
					continue
				}
				subnetIdentifier := capxv1.NutanixResourceIdentifier{}
				if subnet.Type == v1alpha1.NutanixIdentifierName {
					subnetIdentifier.Type = capxv1.NutanixIdentifierName
					if subnet.Name == nil || *subnet.Name == "" {
						continue
					}
					subnetIdentifier.Name = subnet.Name
				} else {
					subnetIdentifier.Type = capxv1.NutanixIdentifierUUID
					if subnet.UUID == nil || *subnet.UUID == "" {
						continue
					}
					subnetIdentifier.UUID = subnet.UUID
				}
				obj.Spec.Template.Spec.Subnets = append(spec.Subnets, subnetIdentifier)
			}
```

File: pkg/handlers/nutanix/mutation/machinedetails/variables_test.go [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+		ptr.To(v1alpha1.ClusterConfigSpec{Nutanix: &v1alpha1.NutanixSpec{}}.VariableSchema()),
+		true,
+		nutanixclusterconfig.NewVariable,
+		capitest.VariableTestDef{
+			Name: "machine details",
+			Vals: v1alpha1.ClusterConfigSpec{
+				ControlPlane: &v1alpha1.NodeConfigSpec{
+					Nutanix: &v1alpha1.NutanixNodeSpec{
+						MachineDetails: &v1alpha1.NutanixMachineDetails{
+							BootType:       "legacy",
```
Comment (dkoshkin) [inline pkg/handlers/nutanix/mutation/machinedetails/variables_test.go:33]:
```suggestion
							BootType:       v1alpha1.NutanixBootTypeLegacy,
```

File: pkg/handlers/nutanix/mutation/prismcentralendpoint/inject.go [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+			obj.Spec.Template.Spec.PrismCentral.Port = prismCentralEndpointVar.Port
+			obj.Spec.Template.Spec.PrismCentral.Insecure = prismCentralEndpointVar.Insecure
+			obj.Spec.Template.Spec.PrismCentral.AdditionalTrustBundle = &credentials.NutanixTrustBundleReference{
+				Kind: credentials.NutanixTrustBundleKindConfigMap,
+				Name: prismCentralEndpointVar.AdditionalTrustBundle,
+			}
+			obj.Spec.Template.Spec.PrismCentral.CredentialRef = &credentials.NutanixCredentialReference{
+				Kind: credentials.SecretKind,
+				Name: prismCentralEndpointVar.CredentialSecret,
+			}
```
Comment (dkoshkin) [inline pkg/handlers/nutanix/mutation/prismcentralendpoint/inject.go:108]: WDYT?
```suggestion
			obj.Spec.Template.Spec.PrismCentral = &credentials.NutanixPrismEndpoint{
				Address:  prismCentralEndpointVar.Host,
				Port:     prismCentralEndpointVar.Port,
				Insecure: prismCentralEndpointVar.Insecure,
				AdditionalTrustBundle: &credentials.NutanixTrustBundleReference{
					Kind: credentials.NutanixTrustBundleKindConfigMap,
					Name: prismCentralEndpointVar.AdditionalTrustBundle,
				},
				CredentialRef: &credentials.NutanixCredentialReference{
					Kind: credentials.SecretKind,
					Name: prismCentralEndpointVar.CredentialSecret,
				},
			}
```

File: hack/examples/bases/nutanix/cluster/kustomization.yaml.tmpl [hunk 10]
```diff
@@ -5,7 +5,7 @@ apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
 
 resources:
-- https://raw.githubusercontent.com/nutanix-cloud-native/cluster-api-provider-nutanix/jira/KRBN-8094/templates/cluster-template-topology.yaml
+- https://raw.githubusercontent.com/nutanix-cloud-native/cluster-api-provider-nutanix/main/templates/cluster-template-topology.yaml
```
Comment (dkoshkin) [inline hack/examples/bases/nutanix/cluster/kustomization.yaml.tmpl:8]: This is reverting from `main` maybe you can drop this commit and rebase instead?

File: hack/examples/bases/nutanix/clusterclass/kustomization.yaml.tmpl [hunk 11]
```diff
@@ -29,3 +29,24 @@ patches:
       path: "/spec/patches"
     - op: "remove"
       path: "/spec/variables"
+# TODO: Temp added patches for NutanixClusterTemplate
```
[the remark this answers is not in the document]
Comment (deepakm-ntnx) [reply hack/examples/bases/nutanix/clusterclass/kustomization.yaml.tmpl:32]: https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix/pull/400 is yet to be merged. will update capx thru that PR and once merged, we can remove it later.

File: pkg/handlers/nutanix/mutation/machinedetails/inject_control_plane.go [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+						continue
+					}
+					subnet.UUID = subnetCRE.UUID
+				}
+				obj.Spec.Template.Spec.Subnets = append(obj.Spec.Template.Spec.Subnets, subnet)
+			}
+			// TODO:deepakm-ntnx assign user provided values
+			obj.Spec.Template.Spec.Project = nil
+			obj.Spec.Template.Spec.AdditionalCategories = []capxv1.NutanixCategoryIdentifier{}
+			obj.Spec.Template.Spec.GPUs = []capxv1.NutanixGPU{}
```
[the remark this answers is not in the document]
Comment (deepakm-ntnx) [reply pkg/handlers/nutanix/mutation/machinedetails/inject_control_plane.go:155]: this is required else was getting some validation errors. will be fixing it in next PR

File: api/v1alpha1/nutanix_node_types.go [hunk 13]
```diff
@@ -0,0 +1,286 @@
...
+	// cluster is to identify the cluster (the Prism Element under management
+	// of the Prism Central), in which the Machine's VM will be created.
+	// The cluster identifier (uuid or name) can be obtained from the Prism Central console
+	// or using the prism_central API.
+	Cluster NutanixResourceIdentifier `json:"cluster"`
+
+	// subnet is to identify the cluster's network subnet to use for the Machine's VM
+	// The cluster identifier (uuid or name) can be obtained from the Prism Central console
+	// or using the prism_central API.
+	Subnets NutanixResourceIdentifiers `json:"subnet"`
```
Comment (dkoshkin) [inline api/v1alpha1/nutanix_node_types.go:96]: This is of type `NutanixResourceIdentifiers` which is already a slice but the `json` representation is `subnet, should this be a single subnet or a list?

File: pkg/handlers/nutanix/mutation/machinedetails/inject_worker.go (hunk 6 above)
Comment (deepakm-ntnx) [reply pkg/handlers/nutanix/mutation/machinedetails/inject_worker.go]: i tried that earlier but we want user to enter it as a string and then we convert that to resource.

**Review by dkoshkin — DISMISSED** [review]
I'm happy with this, thanks @deepakm-ntnx!

Tested locally with both Calico and Cilium, everything looks great.

```
CONTROL_PLANE_ENDPOINT_IP=<>
CONTROL_PLANE_ENDPOINT_PORT=6443

NUTANIX_ENDPOINT=<>
NUTANIX_USER=<>
NUTANIX_PASSWORD=<>
NUTANIX_PRISM_ELEMENT_CLUSTER_NAME=<>
NUTANIX_INSECURE=false
NUTANIX_SUBNET_NAME=<>
NUTANIX_MACHINE_TEMPLATE_IMAGE_NAME=ubuntu-2204-kube-v1.28.7.qcow2

CLUSTER_NAME=dkoshkin-nutanix-caren-1 
CLUSTER_FILE=examples/capi-quick-start/nutanix-cluster-calico-crs.yaml
KUBERNETES_VERSION=v1.28.7 

clusterctl generate cluster ${CLUSTER_NAME} \
  --from ${CLUSTER_FILE} \
  --kubernetes-version ${KUBERNETES_VERSION} \
  --worker-machine-count 1 | \
  kubectl apply --server-side -f -
```

Nodes came up:
```  
$ kubectl get nodes
NAME                                              STATUS   ROLES           AGE     VERSION
dkoshkin-nutanix-caren-1-drvx2-tk7c5              Ready    control-plane   3m13s   v1.28.7
dkoshkin-nutanix-caren-1-md-0-9pmr9-bj6rx-gx982   Ready    <none>          2m15s   v1.28.7

```

**Review by dkoshkin — APPROVED** [review]

**Review by thunderboltsid — APPROVED** [review]
